src/farming/library.cairo

```diff
@@ -0,0 +1,965 @@
+        let (contract_address) = CarbonableFarming_carbonable_project_address_.read();
+        let (slot) = CarbonableFarming_carbonable_project_slot_.read();
+        let (ton_equivalent) = ICarbonableProject.getTonEquivalent(
+            contract_address=contract_address, slot=slot
+        );
+        if (ton_equivalent == 0) {
+            return (sale=0);
+        }
+
+        // [Compute] User sale and convert from g * µ$/t to µ$
+        let (computed) = _compute_user_sale(address=address);
+        let (stored) = CarbonableFarming_sale_.read(address);
+        let (sale_in_dollars, _) = unsigned_div_rem(computed + stored, ton_equivalent);
+
+        return (sale=sale_in_dollars);
+    }
+
+    func current_price{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (
+        current_price: felt
+    ) {
+        alloc_locals;
+
+        let (len, times, _, updated_prices, _) = CarbonableFarming._compute_cumsales();
+        let (time) = get_block_timestamp();
+        let current_price = Math.interpolate(
+            x=time,
+            len=len,
+            xs=times,
+            ys=updated_prices,
+            interpolation=CONST_LEFT,
+            extrapolation=NULL,
+        );
+
+        return (current_price=current_price);
+    }
+
+    func prices{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (
+        len: felt, times: felt*, prices: felt*, updated_prices: felt*, cumsales: felt*
+    ) {
+        alloc_locals;
+
+        return CarbonableFarming._compute_cumsales();
+    }
+
+    //
+    // Externals
+    //
+
+    func set_prices{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        times_len: felt, times: felt*, prices_len, prices: felt*
+    ) {
+        alloc_locals;
+
+        Array.store(key=TIME_SK, len=times_len, values=times);
+        Array.store(key=PRICE_SK, len=prices_len, values=prices);
+
+        return ();
+    }
+
+    func add_price{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        time: felt, price: felt
+    ) {
+        alloc_locals;
+
+        // [Check] Time is later than the project start time
+        let (contract_address) = CarbonableFarming_carbonable_project_address_.read();
+        let (slot) = CarbonableFarming_carbonable_project_slot_.read();
+        let (first_time) = ICarbonableProject.getStartTime(
+            contract_address=contract_address, slot=slot
+        );
+        with_attr error_message("CarbonableFarming: time is sooner than the project start time") {
+            assert_le(first_time + 1, time);
+        }
+
+        // [Check] First time to store
+        let (len) = Array.read_len(key=TIME_SK);
+        if (len == 0) {
+            // [Effect] Update storage
+            Array.add(key=TIME_SK, value=first_time);
+            Array.add(key=TIME_SK, value=time);
+            Array.add(key=PRICE_SK, value=price);
+            Array.add(key=PRICE_SK, value=price);
```

Comment:
this line is redundant

Reply:
Actually this is to handle the first time a price is added which will be to add the same price twice at first time and the specified time.